### ISSUE TITLE
"PC-relative offset overflow in PLT entry" while linking libtorch_cuda.so

### ISSUE THREAD

**Issue by danpetry** [pytorch/pytorch#110986]
### 🐛 Describe the bug

While building [this](https://github.com/AnacondaRecipes/pytorch-feedstock/tree/update-2.1.0-gpu-linux64-debugging) recipe for linux-64, GPU, cuda v11.8, I get the following error:

```
lib/libtorch_cuda.so: PC-relative offset overflow in PLT entry for `_ZN2at6native15gpu_kernel_implI18__nv_hdl_wrapper_tILb0ELb0E11__nv_dl_tagIPFvRNS_6TensorERKS4_S7_fEXadL_ZNS0_43_GLOBAL__N__4fe90af2_10_Dropout_cu_f338191119masked_scale_kernelIbN3c104HalfEfEEvS5_S7_S7_T1_EELj1EEFSD_SD_bEJfEEEEvRNS_18TensorIteratorBaseERKT_'
  collect2: error: ld returned 1 exit status
```

The full log is here: [2buildsplitcuda.log](https://github.com/pytorch/pytorch/files/12861957/2buildsplitcuda.log)

I have already tried adding the `--no-relax` linker flag and `BUILD_SPLIT_CUDA` env variable. The linker flag solved a related problem earlier on in the build (the linker explicitly told me to use this flag) and the build progressed to a later stage, but then I got the above failure and the env variable doesn't seem to be helping.

To reproduce, do this on a machine with cuda 11.8 installed, in a fresh conda environment with conda-build installed (AFAIU you don't need the nvidia hardware):
```
git clone https://github.com/AnacondaRecipes/aggregate.git
cd aggregate
git submodule update --init pytorch-feedstock
cd pytorch-feedstock
git fetch --all
git checkout update-2.1.0-gpu-linux64-debugging
cd ../
conda build pytorch-feedstock
```

### Versions

PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: CentOS Linux 7 (Core) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.17

Python version: 3.9.18 (main, Sep 11 2023, 13:41:44)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.1.55-75.123.amzn2023.x86_64-x86_64-with-glibc2.17
Is CUDA available: N/A
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: Tesla M60
Nvidia driver version: 520.61.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                4
On-line CPU(s) list:   0-3
Thread(s) per core:    2
Core(s) per socket:    2
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 79
Model name:            Intel(R) Xeon(R) CPU E5-2686 v4 @ 2.30GHz
Stepping:              1
CPU MHz:               2700.107
CPU max MHz:           3000.0000
CPU min MHz:           1200.0000
BogoMIPS:              4600.03
Hypervisor vendor:     Xen
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              256K
L3 cache:              46080K
NUMA node0 CPU(s):     0-3
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single pti fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx xsaveopt

Versions of relevant libraries:
[pip3] No relevant packages
[conda] No relevant packages

cc @malfet @seemethere

**Comment by danpetry** [comment]
After comparison with the log output in the pytorch CI, I've taken the following steps re changing environment variables and compiler/linker flags, but haven't managed to solve the problem. Some are more relevant than others, but I'm including all changes here in case one's more relevant than I think:
- included -fvisibility-inlines-hidden, -fno-plt, -ffunction-sections, -fabi-version=11 compiler flags
- included --no-relax, -z,now (was -z,lazy), --as-needed, -dead_strip_dylibs, linker flags
- changed gcc version from 11.2.0 to 9.3.0
- removed `TORCH_NVCC_FLAGS="-Xfatbin -compress-all"`, `NCCL_ROOT_DIR=/usr/local/cuda`, `USE_STATIC_NCCL=1`, `BUILD_SPLIT_CUDA="ON"` env variables
- added `TH_BINARY_BUILD=1`, `INSTALL_TEST=0`, `USE_STATIC_CUDNN=0`, `_GLIBCXX_USE_CXX11_ABI=0` env variables

I'm running out of ideas, so if anyone's got time, some eyes on the issue would be much appreciated.

**Comment by danpetry** [comment]
I solved this. The issue was that the `TORCH_CUDA_ARCH_LIST` I was using was very long, increasing the binary size beyond what the linker could handle. I put a comprehensive list of architectures in the list, not realising that the `PTX` entry allowed compatibility with any archs not listed (albeit with a performance hit).

 I set to to the same as pytorch's build system:

`Texport TORCH_CUDA_ARCH_LIST="3.5+PTX;5.0;6.0;6.1;7.0;7.5;8.0;8.6;9.0"`
